### PR TITLE
Add the capability to use cached meshes and ICs

### LINKED ISSUE #175: Add ability to run tests from cached initial conditions
Right now, any tests run in compass must first generate their own mesh and initial condition files. This is fine for QU240, but for other global resolutions the init step can be lengthy, so it is more practical to point to cached initial condition files on disk. We could have a cached location on each supported machine (most already do), and compass could download files if they are not already there.  We could point to E3SM files already here:
https://web.lcrc.anl.gov/public/e3sm/inputdata/ocn/mpas-o/
or add files here:
https://web.lcrc.anl.gov/public/e3sm/mpas_standalonedata/mpas-ocean/

My motivation here is performance testing at varying resolutions, but this could be useful for any test or stand-alone simulation at resolutions higher that QU240.

### PR DESCRIPTION
**Pull request by xylar**
This merge adds a new capability for steps to have cached outputs in a `compass_cache` database.  Files in this database have a directory structure similar to the work directory (but without the MPAS core subdirectory, which is redundant).  The files include a date stamp so that new revisions can be added without removing older ones (supported by older `compass` versions).

As an example, here are cached files for the `mesh` and `init` test cases for the `QU240` mesh:
```
compass_cache/global_ocean/QU240/mesh/mesh/critical_passages_mask_final.210727.nc
compass_cache/global_ocean/QU240/mesh/mesh/culled_graph.210727.info
compass_cache/global_ocean/QU240/mesh/mesh/culled_mesh.210727.nc
compass_cache/global_ocean/QU240/PHC/init/initial_state/init_mode_forcing_data.210727.nc
compass_cache/global_ocean/QU240/PHC/init/initial_state/initial_state.210727.nc
```

The mapping between outputs in the cached versions of steps and the files in the `compass_cache` database are maintained in a file in `compass.<mpas_core>` called `cached_files.json`.  This file contains a python dictionary that maps from the output files in the cached versions of each step (the symlinks) to those in the database (the targets).  For example:

``` json

{
    "ocean/global_ocean/cached/QU240/mesh/mesh/culled_mesh.nc": "global_ocean/QU240/mesh/mesh/culled_mesh.210727.nc",
    "ocean/global_ocean/cached/QU240/mesh/mesh/culled_graph.info": "global_ocean/QU240/mesh/mesh/culled_graph.210727.info",
    "ocean/global_ocean/cached/QU240/mesh/mesh/critical_passages_mask_final.nc": "global_ocean/QU240/mesh/mesh/critical_passages_mask_final.210727.nc",
    "ocean/global_ocean/cached/QU240/PHC/init/initial_state/initial_state.nc": "global_ocean/QU240/PHC/init/initial_state/initial_state.210727.nc",
    "ocean/global_ocean/cached/QU240/PHC/init/initial_state/init_mode_forcing_data.nc": "global_ocean/QU240/PHC/init/initial_state/init_mode_forcing_data.210727.nc"
}
```

A new command, `compass cache` has been added to aid in updating `cached_files.json`.  This command is only available on Anvil and Chrysalis, since you can only copy files from a compass work directory onto the LCRC server from these two machines.  You run `compass cache` from the base work directory, giving the relative paths to the step(s) that you want to cache output files from.  For example:
```
compass cache -i ocean/global_ocean/QU240/mesh/mesh 
```
This will:
1. copy the output files from the `ocean/global_ocean/QU240/mesh/mesh` step into the appropriate `compass_cache` location on the LCRC server and
2. add these files to a local `ocean_cached_files.json` that can then be copied to `compass/<mpas_core>/cached_files.json` in a local compass branch so it is ready for a PR.

If you want, you can provide several steps with the `-i` flag or just call `compass cache` several times.  In either case, each call will update the local `ocean_cached_files.json`.

See the design doc and updated documentation on details about setting up test cases and suites with cached outputs.

closes #175

### CODE REVIEW

**Comment by xylar** [comment]
@mark-petersen, it seems that chrysalis still isn't working (you can log in but there aren't any useful directories available).  I'll upload QU240 and QUwISC240 cached files as soon as it comes back.  Other resolutions are too big to upload from my laptop so I'll create them on Chrysalis directly when I get a chance.

I tested the performance tests with QU240 and QUwISC240 and they worked as expected (the mesh and init test cases did nothing, as expected).  I didn't test restart, decomposition or thread tests yet but I think they'll work fine if performance did.

Let me know if you have concerns.

Obviously, this needs to be added to the documentation.  I'll do that when I get back before this gets merged.

**Comment by xylar** [comment]
If you want to test in the meantime, here's a file you can untar in your local `mpas-ocean` directory that should currently contain `mesh_database`, `bathymetry_database` and `initial_condition_database`: https://drive.google.com/file/d/1QmSO6Q_l8ngxV6FKkEPMGoHKk_FAgXST/view?usp=sharing

**Comment by mark-petersen** [comment]
@xylar, thanks, this is great! I'm chairing sessions at SIAM today and tomorrow, will work on this next week.

**Comment by xylar** [comment]
@mark-petersen, I'm working on this.  I gave it some thought today.  It needs to support date stamps for the cached files, which is not the case with the implementation here.  It also could be made much more general, allowing any step from any core to have a cached version, without much trouble.  So I will be working on that on the train today.  It's about 1/2 done already.

**Comment by xylar** [comment]
## Testing

I have tested all cached steps with 5 meshes on Chrysalis with Intel compilers and Intel MPI.  All worked as expected (cached `mesh`, `initial_state` and `ssh_adjustment` steps did nothing, and forward tests ran as expected and verification was successful.

I will also test the QU*240 mesh tests under Linux to ensure that downloading of the files from the LCRC server (which doesn't happen on Chrysalis, since they are local) also works as expected.  I will update this comment when I have done this testing.

**Comment by xylar** [comment]
@mark-petersen, this is ready to test for all but the SOwISC12to60 test case.  I want to get #187 in first, then rebase this and produce the cache files for that mesh as well.  But feel free to test and review in the meantime.

**Comment by xylar** [comment]
@mark-petersen, sorry for being a moving target but in the process of writing the documentation for this PR, I realize I'm not happy with the current implementation either.  It's very clumsy to have "normal" and "cached" versions of steps (and test cases).  Instead, it would make sense for *every* step to be cached with the same workflow.  There should be relatively simple way to specify which steps are cached and which are not.  So far, the only reasonable way I can come up with to do that is with different test suites.

I would benefit from having a brief chat whenever you're free.  Based on that chat, it might be worth putting this PR on hold and making a proper design doc for this addition.  That would give @matthewhoffman (and others who might be interested) a chance to weigh in at the expense of taking quite a bit longer to get finalized and merged.  Given that this functionality could have a lot of different uses if it's done properly, I think it's likely worth doing a proper design doc.

**Comment by xylar** [comment]
@mark-petersen, I have finished my testing and this is ready for final review, along with #189.

**Comment by xylar** [comment]
@matthewhoffman, it might be best to hold off until we know whether @mark-petersen has any changes he is requesting.  Before the final merge, I would like to have you run your test suites to make sure nothing gets unexpectedly affected by these changes.  I certainly don't expect any changes if you're not using caching.

**Comment by xylar** [comment]
@mark-petersen, I've rebased this an handled the conflicts with #191.  I ran a few tests and I don't think the rebase introduced any new issues.  I am able to create a custom test suite that includes steps with cached outputs, which is what this branch combined with #191 should allow.

**Comment by mark-petersen** [comment]
@xylar, this looks great. I am running cases on grizzly and chrysalis.

I set up a test suite on grizzly like this:
```
compass suite -s -c ocean -t nightly -w $n/210809_pr_cache
	Setting up test cases:
...
  ocean/global_ocean/QUwISC240/mesh
    steps with cached outputs: mesh
Downloading culled_mesh.210803.nc (8.8MiB)...
100% |##################################################################################################| Time:  0:00:00
  culled_mesh.210803.nc done.
Downloading culled_graph.210803.info (207.0KiB)...
100% |##################################################################################################| Time:  0:00:00
  culled_graph.210803.info done.
... etc
```
and indeed it downloaded a bunch of files with dates (very nice!) and was able to run successfully on a compute node, but it looks like those compute the initial condition, rather than use the cached files. 

My first question is, where are the files stored on a particular machine? That information is not clear from the screen output. I was able to find it in the files by grepping:
```
compass/machines/chrysalis.cfg:8:ocean_database_root = /lcrc/group/e3sm/public_html/mpas_standalonedata/mpas-ocean
```
but maybe we should make that location more obvious. Since the downloads only occur the first time, we could add
```
Downloading initial_state.210803.nc (23.7MiB) to /lcrc/group/e3sm/public_html/mpas_standalonedata/mpas-ocean...
```
or even the full path, since it is handy there.

Moving on, I tried this:
```
compass list|grep PHC
  41: ocean/global_ocean/QU240/PHC/init
  42: ocean/global_ocean/QU240/PHC/performance_test
```
and tried cache for 41, to run 42:
```
compass setup -w $n/210809_pr_cache_4 -n 41c 42
Setting up test cases:
  ocean/global_ocean/QU240/PHC/init
    steps with cached outputs: initial_state
Downloading init_mode_forcing_data.210803.nc (529.1KiB)...
100% |##################################################################################################| Time:  0:00:00
  init_mode_forcing_data.210803.nc done.
  ocean/global_ocean/QU240/PHC/performance_test
target cores: 4
minimum cores: 1
```
so the setup works. But the run fails:
```
compass run
ocean/global_ocean/QU240/PHC/init
  test execution:      SUCCESS
ocean/global_ocean/QU240/PHC/performance_test
  test execution:      ERROR
  see: case_outputs/ocean_global_ocean_QU240_PHC_performance_test.log

tail -n 1 case_outputs/ocean_global_ocean_QU240_PHC_performance_test.log
OSError: input file(s) missing in step forward of ocean/global_ocean/QU240/PHC/performance_test: ['/lustre/scratch3/turquoise/mpeterse/runs/n/210809_pr_cache_4/ocean/global_ocean/QU240/mesh/mesh/culled_graph.info']
```
I got the same thing with QU240wISC
```
compass setup -w $n/210809_pr_cache_5 -n 61c 62
Setting up test cases:
  ocean/global_ocean/QUwISC240/PHC/init
    steps with cached outputs: initial_state ssh_adjustment
  ocean/global_ocean/QUwISC240/PHC/performance_test
...
cd $n/210809_pr_cache_5
(dev_compass_1.0.0) gr0655:210809_pr_cache_5$ compass run
ocean/global_ocean/QUwISC240/PHC/init
  test execution:      SUCCESS
ocean/global_ocean/QUwISC240/PHC/performance_test
  test execution:      ERROR
  see: case_outputs/ocean_global_ocean_QUwISC240_PHC_performance_test.log
...
tail -n 1 case_outputs/ocean_global_ocean_QUwISC240_PHC_performance_test.log
OSError: input file(s) missing in step forward of ocean/global_ocean/QUwISC240/PHC/performance_test: ['/lustre/scratch3/turquoise/mpeterse/runs/n/210809_pr_cache_5/ocean/global_ocean/QUwISC240/mesh/mesh/culled_graph.info']
```

It looks like the cache functionality might work, but it is just missing some graph.info files on the server. Is there a case that worked successfully from remotely cached files for you? Thanks.

**Comment by xylar** [comment]
> My first question is, where are the files stored on a particular machine?

I think this is documented but the base is in `compass/machines/grizzly.cfg` and similarly for other machines.  

> maybe we should make that location more obvious. Since the downloads only occur the first time, we could add
> Downloading initial_state.210803.nc (23.7MiB) to /lcrc/group/e3sm/public_html/mpas_standalonedata/mpas-ocean...
> or even the full path, since it is handy there.

I could add that.  I guess I was hesitating to clutter things up but it could be helpful.

**Comment by xylar** [comment]
> Moving on, I tried this:
> compass list|grep PHC
>  41: ocean/global_ocean/QU240/PHC/init
>  42: ocean/global_ocean/QU240/PHC/performance_test

You are missing the `mesh` step for each of these runs.  In this case, you want the cached version because it would be silly to run with the cached version of the init step but computing the mesh.  The thing is that the `performance_test` depends on not just the `init` step but also directly on the outputs from the `mesh` step.  You have to have both.

**Comment by mark-petersen** [comment]
> You are missing the mesh step for each of these runs

Great!, yes, silly me. I see that now. This works now:
```
compass setup -w $n/210809_pr_cache_7 -n 60c 61c 62
Setting up test cases:
  ocean/global_ocean/QUwISC240/mesh
    steps with cached outputs: mesh
  ocean/global_ocean/QUwISC240/PHC/init
    steps with cached outputs: initial_state ssh_adjustment
  ocean/global_ocean/QUwISC240/PHC/performance_test
...
00:00 PASS ocean_global_ocean_QUwISC240_mesh
00:00 PASS ocean_global_ocean_QUwISC240_PHC_init
00:28 PASS ocean_global_ocean_QUwISC240_PHC_performance_test
```
for some reason this one dies in the performance forward step:
```
compass setup -w $n/210809_pr_cache_8 -n 40c 41c 42
Setting up test cases:
  ocean/global_ocean/QU240/mesh
    steps with cached outputs: mesh
  ocean/global_ocean/QU240/PHC/init
    steps with cached outputs: initial_state
  ocean/global_ocean/QU240/PHC/performance_test
```
during bootstrapping reading `init.nc` with a PIO2 error on grizzly and gnu (but it worked for QUwISC240). Can you try that sequence and see if it works for you?

If we change the init steps to copy the graph.info from the mesh directory (or reproduce it) rather than link to it, then we only need the init step, not the mesh step as well. I think that is preferable. Simulation steps don't link to any netcdf files in the mesh step, so it seems silly to require the cached mesh step with potentially large netcdf files, just for a graph.info file.

**Comment by xylar** [comment]
@mark-petersen

> If we change the init steps to copy the graph.info from the mesh directory (or reproduce it) rather than link to it, then we only need the init step, not the mesh step as well. I think that is preferable. Simulation steps don't link to any netcdf files in the mesh step, so it seems silly to require the cached mesh step with potentially large netcdf files, just for a graph.info file.

I don't love that design personally.  The `initial_state` step doesn't need the `graph.info` file so it seems weird to me that it should have it as both an input and an output just because it happens to be an intermediate step between the `mesh` and various forward steps.

An alternative is for the various forward steps to just generate the graph file on the fly as they need it.  We have a python function to generate it so there's no need to take it from the culled mesh step.  What do you think of that solution.

**Comment by mark-petersen** [comment]
My highest priority is speed, so I would not want to add an extra computation to every forward step to reproduce the graph.info file every time. I'd prefer to have the init step be a 'one stop shop' for the init, forcing, and graph file. Then we only have a single cached case to grab for a resolution:
```
compass setup -w $n/210809_pr_cache_10 -n 61c 62
Setting up test cases:
  ocean/global_ocean/QUwISC240/PHC/init
    steps with cached outputs: initial_state ssh_adjustment
  ocean/global_ocean/QUwISC240/PHC/performance_test
```
That is really elegant, and easy to remember.

**Comment by xylar** [comment]
Okay, I don't agree with the elegant part but I'll do that if that's what you want.

**Comment by mark-petersen** [comment]
Yes, thanks. 'Eye of the beholder' and all that. But it does reduce the download requirements from the mesh step, and that will save some time for our larger meshes.

**Comment by xylar** [comment]
> Yes, thanks. 'Eye of the beholder' and all that. But it does reduce the download requirements from the mesh step, and that will save some time for our larger meshes.

But that's a one-time download on supported machines.  And the mesh is available if you want to run the mesh step as cached and the init step not.  You are treating this as if your current use case is going to be the primary one but I don't know that I agree with that assessment.  The changes you want are going to potentially cause a loss of the generality of the implementation.

**Comment by xylar** [comment]
Anyway, it's pretty late here and I won't have time to implement this change tomorrow, so I guess that means this won't get addressed until late this week.

**Comment by mark-petersen** [comment]
OK. Thanks for all your work so far. This is a fantastic new feature.

**Comment by xylar** [comment]
> for some reason this one dies in the performance forward step:
> compass setup -w $n/210809_pr_cache_8 -n 40c 41c 42

@mark-petersen, it looks like you must have interrupted the downloading of the file so it's broken:

```
$ ls -lah /usr/projects/regionalclimate/COMMON_MPAS/ocean/grids/compass_cache/global_ocean/QU240/PHC/init/initial_state/initial_state.210803.nc
-rw-r--r-- 1 mpeterse mpeterse 2.7M Aug  9 08:31 /usr/projects/regionalclimate/COMMON_MPAS/ocean/grids/compass_cache/global_ocean/QU240/PHC/init/initial_state/initial_state.210803.nc
```
but the same file on the LCRC server is 23 MB:
https://web.lcrc.anl.gov/public/e3sm/mpas_standalonedata/mpas-ocean/compass_cache/global_ocean/QU240/PHC/init/initial_state/

There are 2 options to fix this.  The first is just to delete the file (which only you have write permission to) and rerun the test case.  The other is to make a user config file and set:
```
# Options related to downloading files
[download]

# whether to check the size of files that have been downloaded to make sure
# they are the right size
check_size = True
```
We don't have this as the default because it is very time consuming to check if each downloaded file is the right size each time a test case gets set up.  But if you know you're downloading files and want to make sure they're complete, this is how.

By the way, you have also half-downloaded files on Cori in the past, causing us some hassle with the legacy COMPASS.  If you're downloading files into a shared space, it's really important to let the downloads complete, or to carefully clean up if you don't.  (I guess that's partly why you were trying to figure out where the files end up?)

**Comment by xylar** [comment]
@mark-petersen, it was pretty easy to implement what you wanted in terms of the `graph.info` in #203.  I rebased this on that branch, created new cache files for all test cases except `SOwIsC12to60` (still running), and updated the list of cache files.

Please see if you can delete and re-download the cache files on Grizzly.  See if you can run without the `mesh` steps this time.  It worked for me.

I'll force-push to update my last commit once `SOwISC12to60` results are there for caching.

File: compass/ocean/cached_files.json
Comment (xylar) [inline compass/ocean/cached_files.json:39]: These still need to be updated following #203.

**Comment by xylar** [comment]
@mark-petersen, this is ready for you to re-review along with #203

**Review by mark-petersen — APPROVED** [review]
I tested performance tests using cached init on grizzly and cori for QU240, QUwISC240, and EC60to30. Everything works great! See details here: https://github.com/MPAS-Dev/compass/pull/203#pullrequestreview-726426422

**Comment by mark-petersen** [comment]
Sorry about the half-downloads. I had no idea, but I'll watch for it in the future.

**Comment by xylar** [comment]
@mark-petersen, thanks again.  I rebased after merging #203 and will merge this once tests have passed.

**Comment by xylar** [comment]
Oh, sorry.  @matthewhoffman, did you want to do a test run with this branch?  It affects the general framework but I'm confident it shouldn't have any unexpected impacts on `landice`.  Still, doesn't hurt to check...

**Comment by matthewhoffman** [comment]
@xylar , I will do that today - thanks.

**Comment by matthewhoffman** [comment]
@xylar , I got some unexpected differences against the baseline yesterday.  I suspect they may be due to an outdated baseline and not a problem with this PR.  I need to redo my test more carefully today (make sure my baseline is up to date with master, and then re-test this branch).  I'm starting that now.

**Comment by matthewhoffman** [comment]
My corrected test passed.  Go ahead and merge.

**Comment by xylar** [comment]
Thanks very much, @matthewhoffman!

**Pull request merged**
